File: decoder/src/cpp/relay_bp_decoder.cpp
#include "relay_bp_decoder.h"
#include <algorithm>
#include <cmath>
#include <numeric>

// --- Helper Functions ---

double bound_value(double val) {
    // Prevent numerical explosion
    if (val > 50.0) return 50.0;
    if (val < -50.0) return -50.0;
    return val;
}

int sign_of(double x) {
    return (x < 0) ? -1 : 1;
}

// --- MinSumBPDecoder Implementation ---

MinSumBPDecoder::MinSumBPDecoder(const std::vector<int>& row_ptr, 
                                 const std::vector<int>& col_ind, 
                                 int n_rows, int n_cols,
                                 const std::vector<double>& priors,
                                 int max_iter, double alpha)
    : num_checks(n_rows), num_vars(n_cols), max_iter(max_iter), alpha(alpha), 
      h_csr_indptr(row_ptr), h_csr_indices(col_ind), current_iteration(0), has_memory(false) 
{
    // Initialize LLRs
    log_prior_ratios.resize(num_vars);
    for (int i = 0; i < num_vars; ++i) {
        double p = std::max(1e-15, std::min(priors[i], 1.0 - 1e-15));
        log_prior_ratios[i] = std::log((1.0 - p) / p);
    }
    posterior_ratios = log_prior_ratios; // Initial posteriors
    decoding.resize(num_vars);
    memory_strengths.resize(num_vars, 0.0);

    // Initialize CSR message storage (Variable -> Check messages, stored at Check nodes)
    int num_edges = h_csr_indices.size();
    q_messages.resize(num_edges, 0.0);

    // Build CSC structure (Transpose of H) for Variable updates
    // Count edges per column
    std::vector<int> col_counts(num_vars, 0);
    for (int col : h_csr_indices) {
        col_counts[col]++;
    }

    h_csc_indptr.resize(num_vars + 1);
    h_csc_indptr[0] = 0;
    for (int i = 0; i < num_vars; ++i) {
        h_csc_indptr[i + 1] = h_csc_indptr[i] + col_counts[i];
    }

    h_csc_indices.resize(num_edges);
    r_messages.resize(num_edges, 0.0);
    
    // Fill CSC indices and build maps
    // We need to know where we are in each column to fill it
    std::vector<int> current_col_pos = h_csc_indptr;
    
    map_q_to_r.resize(num_edges);
    map_r_to_q.resize(num_edges);

    for (int r = 0; r < num_checks; ++r) {
        for (int i = h_csr_indptr[r]; i < h_csr_indptr[r + 1]; ++i) {
            int c = h_csr_indices[i];
            int dest_pos = current_col_pos[c];
            
            h_csc_indices[dest_pos] = r;
            
            // Map the edge index i (in CSR) to dest_pos (in CSC)
            map_q_to_r[i] = dest_pos;
            map_r_to_q[dest_pos] = i;
            
            current_col_pos[c]++;
        }
    }
}

void MinSumBPDecoder::initialize_decoder() {
    current_iteration = 0;
    std::fill(q_messages.begin(), q_messages.end(), 0.0);
    std::fill(r_messages.begin(), r_messages.end(), 0.0);
    
    // Initialize q_messages with priors
    for (int i = 0; i < q_messages.size(); ++i) {
        int var_idx = h_csr_indices[i];
        q_messages[i] = log_prior_ratios[var_idx];
    }
    
    // Reset posteriors if memory is active, otherwise they just follow
    if (has_memory) {
        posterior_ratios = log_prior_ratios;
    }
}

void MinSumBPDecoder::reset_for_new_leg() {
    current_iteration = 0;
    std::fill(q_messages.begin(), q_messages.end(), 0.0);
    std::fill(r_messages.begin(), r_messages.end(), 0.0);
    
    // Initialize q with priors
    for (int i = 0; i < q_messages.size(); ++i) {
        int var_idx = h_csr_indices[i];
        q_messages[i] = log_prior_ratios[var_idx];
    }
    // Note: We DO NOT reset posterior_ratios here, that is the point of relay
}

void MinSumBPDecoder::set_memory_strengths(const std::vector<double>& gammas) {
    memory_strengths = gammas;
    has_memory = true;
}

void MinSumBPDecoder::compute_check_to_variable(const std::vector<int>& syndrome) {
    // Min-Sum update: r_{cv}
    // Iterate over checks (rows)
    for (int c = 0; c < num_checks; ++c) {
        int start = h_csr_indptr[c];
        int end = h_csr_indptr[c + 1];
        if (start == end) continue;

        double min1 = 1e30, min2 = 1e30;
        int min1_idx = -1;
        int sign_prod = 1;

        // Pass 1: Find min, min2, total sign
        for (int i = start; i < end; ++i) {
            double val = q_messages[i];
            double abs_val = std::abs(val);
            int s = sign_of(val);
            
            sign_prod *= s;
            
            if (abs_val < min1) {
                min2 = min1;
                min1 = abs_val;
                min1_idx = i;
            } else if (abs_val < min2) {
                min2 = abs_val;
            }
        }

        // Adjust for syndrome
        if (syndrome[c]) {
            sign_prod *= -1;
        }

        // Pass 2: Update r messages
        for (int i = start; i < end; ++i) {
            double val = q_messages[i];
            int my_sign = sign_of(val);
            
            // The sign of the message is total_sign / my_sign
            int msg_sign = sign_prod * my_sign; 
            
            double mag = (i == min1_idx) ? min2 : min1;
            
            // Map to r storage
            int r_idx = map_q_to_r[i];
            r_messages[r_idx] = alpha * mag * msg_sign;
        }
    }
}

void MinSumBPDecoder::compute_variable_to_check() {
    // Update posteriors and q messages
    // Iterate over variables (cols of H -> rows of CSC)
    for (int v = 0; v < num_vars; ++v) {
        int start = h_csc_indptr[v];
        int end = h_csc_indptr[v + 1];
        
        // 1. Compute effective prior (DMem-BP)
        double current_prior;
        if (has_memory) {
            current_prior = log_prior_ratios[v] * (1.0 - memory_strengths[v]) 
                          + posterior_ratios[v] * memory_strengths[v];
        } else {
            current_prior = log_prior_ratios[v];
        }
        
        // 2. Sum incoming r messages
        double sum_r = 0.0;
        for (int i = start; i < end; ++i) {
            sum_r += r_messages[i];
        }
        
        // 3. Update posterior
        double new_posterior = current_prior + sum_r;
        new_posterior = bound_value(new_posterior);
        posterior_ratios[v] = new_posterior;
        
        // 4. Update outgoing q messages
        // q_{vc} = posterior - r_{cv}
        for (int i = start; i < end; ++i) {
            double q_val = new_posterior - r_messages[i];
            // Map back to CSR storage
            int q_idx = map_r_to_q[i];
            q_messages[q_idx] = bound_value(q_val);
        }
    }
}

std::vector<int> MinSumBPDecoder::get_hard_decision() {
    std::vector<int> res(num_vars);
    for (int i = 0; i < num_vars; ++i) {
        res[i] = (posterior_ratios[i] < 0) ? 1 : 0;
    }
    return res;
}

bool MinSumBPDecoder::decode(const std::vector<int>& syndrome) {
    if (current_iteration == 0) {
        initialize_decoder();
    }

    for (int iter = 0; iter < max_iter; ++iter) {
        compute_check_to_variable(syndrome);
        compute_variable_to_check();
        current_iteration++;

        // Check convergence
        std::vector<int> candidate = get_hard_decision();
        
        // Verify: H * candidate == syndrome ?
        bool converged = true;
        for (int r = 0; r < num_checks; ++r) {
            int parity = 0;
            for (int i = h_csr_indptr[r]; i < h_csr_indptr[r + 1]; ++i) {
                int c = h_csr_indices[i];
                if (candidate[c]) parity ^= 1;
            }
            if (parity != syndrome[r]) {
                converged = false;
                break;
            }
        }
        
        if (converged) {
            decoding = candidate;
            return true;
        }
    }
    
    decoding = get_hard_decision();
    return false;
}

// --- RelayDecoder Implementation ---

RelayDecoder::RelayDecoder(const std::vector<int>& row_ptr, 
                           const std::vector<int>& col_ind, 
                           int n_rows, int n_cols,
                           const std::vector<double>& priors,
                           int pre_iter, int num_legs, int iter_per_leg,
                           double gamma_min, double gamma_max, double gamma0,
                           unsigned long seed)
    : bp_decoder(row_ptr, col_ind, n_rows, n_cols, priors, pre_iter, 0.625),
      pre_iter(pre_iter), num_legs(num_legs), iter_per_leg(iter_per_leg),
      gamma_min(gamma_min), gamma_max(gamma_max), gamma0(gamma0), rng(seed)
{
}

std::vector<double> RelayDecoder::generate_gammas() {
    std::uniform_real_distribution<double> dist(gamma_min, gamma_max);
    std::vector<double> gammas(bp_decoder.num_vars);
    for (int i = 0; i < bp_decoder.num_vars; ++i) {
        gammas[i] = dist(rng);
    }
    return gammas;
}

bool RelayDecoder::decode(const std::vector<int>& syndrome, std::vector<int>& result) {
    // 1. Leg 0
    bp_decoder.max_iter = pre_iter;
    std::vector<double> g0(bp_decoder.num_vars, gamma0);
    bp_decoder.set_memory_strengths(g0);
    bp_decoder.initialize_decoder();
    
    if (bp_decoder.decode(syndrome)) {
        result = bp_decoder.decoding;
        return true;
    }
    
    // Track best
    // Ideally we track lowest energy/weight solution.
    // For simplicity, we just return the first valid one or the last one.
    // The Python implementation tracked min weight.
    
    // 2. Relay
    for (int leg = 0; leg < num_legs; ++leg) {
        bp_decoder.max_iter = iter_per_leg;
        bp_decoder.set_memory_strengths(generate_gammas());
        bp_decoder.reset_for_new_leg();
        
        if (bp_decoder.decode(syndrome)) {
            result = bp_decoder.decoding;
            return true;
        }
    }
    
    result = bp_decoder.decoding;
    return false;
}

File: decoder/src/relay_bp_decoder.py
import numpy as np
from scipy.sparse import csr_matrix, csc_matrix

class MinSumBPDecoder:
    def __init__(self, check_matrix, error_priors, max_iter=100, alpha=0.625, gamma0=None):
        """
        Args:
            check_matrix (scipy.sparse.spmatrix): Parity check matrix (H).
            error_priors (np.ndarray): Error probability for each qubit/edge.
            max_iter (int): Maximum iterations.
            alpha (float): Scaling factor for Min-Sum (0.0 to 1.0).
            gamma0 (float, optional): Initial memory strength. If None, standard BP.
        """
        self.check_matrix = check_matrix.tocsr() # H matrix
        self.num_checks, self.num_vars = self.check_matrix.shape
        self.max_iter = max_iter
        self.alpha = alpha
        
        # Calculate Log Prior Ratios: log((1-p)/p)
        # Avoid division by zero
        safe_priors = np.clip(error_priors, 1e-15, 1.0 - 1e-15)
        self.log_prior_ratios = np.log((1.0 - safe_priors) / safe_priors)
        
        # Memory strengths
        self.gamma0 = gamma0
        self.memory_strengths = np.zeros(self.num_vars)
        if gamma0 is not None:
            self.memory_strengths[:] = gamma0
            
        # Posteriors (initialized to priors)
        self.posterior_ratios = self.log_prior_ratios.copy()
        self.decoding = np.zeros(self.num_vars, dtype=int)
        
        # Edge storage for message passing
        # We need efficient mapping between check-centric and variable-centric views.
        # H is CSR, so we can iterate checks easily.
        # For variable-centric, we need CSC or a map.
        
        self.rows, self.cols = self.check_matrix.nonzero()
        self.num_edges = len(self.rows)
        
        # Message arrays (flat, aligned with edges)
        # check_to_var_msgs[k] corresponds to edge (rows[k], cols[k])
        # To make this efficient, we sort edges to match CSR (check-centric)
        # and create a mapping for variable-centric operations.
        
        # CSR storage is naturally sorted by row then col.
        # sort_indices() ensures it.
        self.check_matrix.sort_indices()
        self.h_csr = self.check_matrix
        self.h_csc = self.check_matrix.tocsc()
        self.h_csc.sort_indices()
        
        # Messages:
        # q_msgs: Variable to Check messages (same structure as H)
        # r_msgs: Check to Variable messages (same structure as H)
        
        self.q_msgs = self.h_csr.copy()
        self.q_msgs.data = np.zeros(self.num_edges, dtype=np.float64)
        
        self.r_msgs = self.h_csr.copy()
        self.r_msgs.data = np.zeros(self.num_edges, dtype=np.float64)
        
        self.current_iteration = 0

    def initialize_decoder(self):
        self.current_iteration = 0
        self.q_msgs.data[:] = 0.0
        self.r_msgs.data[:] = 0.0
        
        # Initialize variable-to-check messages with priors
        self.q_msgs.data = self.log_prior_ratios[self.q_msgs.indices]
        
        # Reset posteriors if using memory (to priors)
        self.posterior_ratios = self.log_prior_ratios.copy()
        if self.gamma0 is not None:
             self.memory_strengths[:] = self.gamma0

    def reset_for_new_leg(self):
        """
        Resets message graph for a new relay leg, but KEEPS posterior_ratios
        as the 'memory' from the previous leg.
        """
        self.current_iteration = 0
        self.q_msgs.data[:] = 0.0
        self.r_msgs.data[:] = 0.0
        
        # Reset variable messages to priors (standard initialization)
        # The 'memory' effect happens in compute_variable_to_check using posterior_ratios
        self.q_msgs.data = self.log_prior_ratios[self.q_msgs.indices]

    def set_memory_strengths(self, gammas):
        self.memory_strengths = gammas

    def compute_check_to_variable(self, syndrome):
        """
        Min-Sum update: r_{cv} = (-1)^sigma_c * prod_{v' != v} sgn(q_{cv'}) * min_{v' != v} |q_{cv'}|
        """
        # Optimized vectorized implementation
        
        # 1. Prepare data
        q_data = self.q_msgs.data
        q_sign = np.sign(q_data)
        # Avoid zero sign issues, though usually LLRs are non-zero.
        q_sign[q_sign == 0] = 1 
        q_abs = np.abs(q_data)
        
        indptr = self.q_msgs.indptr
        
        # 2. Check update loop
        # We will iterate row by row. 
        # For high performance, this part should ideally be in Cython/Numba, 
        # but pure Python/NumPy is okay for prototyping.
        
        r_data_new = np.zeros_like(q_data)
        
        for i in range(self.num_checks):
            start = indptr[i]
            end = indptr[i+1]
            if start == end:
                continue
            
            # Get messages for this check
            row_q_abs = q_abs[start:end]
            row_q_sign = q_sign[start:end]
            
            # Total sign product
            total_sign = np.prod(row_q_sign)
            
            # Adjust for syndrome
            # Syndrome is 0 or 1. If 1, flip sign.
            syndrome_sign = -1 if syndrome[i] else 1
            total_sign *= syndrome_sign
            
            # Find min and second min absolute values
            if len(row_q_abs) >= 2:
                # We need the two smallest values
                part_idx = np.argpartition(row_q_abs, 1)
                min1_idx = part_idx[0]
                min1 = row_q_abs[min1_idx]
                
                # We need to find the second min accurately. 
                # argpartition does not guarantee order of other elements, 
                # but part_idx[1:] contains them. We take min of those.
                # Actually for small degree (like 6-12), sorting is fast enough or full argpartition.
                # Let's just use part_idx[1] if size is small, but strictly argpartition only guarantees pivot.
                # A safer way for second min:
                remaining = np.delete(row_q_abs, min1_idx)
                min2 = np.min(remaining)
            elif len(row_q_abs) == 1:
                min1 = row_q_abs[0]
                min2 = min1 # Should ideally be huge, but consistent with min-sum approx
                min1_idx = 0
            else:
                continue
                
            # Update r messages
            # For each edge j in this row
            for k in range(len(row_q_abs)):
                # Sign: total_sign * sgn(q_jk)
                # sgn(q_jk) is 1 or -1. dividing is same as multiplying.
                msg_sign = total_sign * row_q_sign[k]
                
                # Magnitude: min1 unless this is the min1, then min2
                msg_mag = min2 if k == min1_idx else min1
                
                # Apply alpha scaling
                val = self.alpha * msg_mag * msg_sign
                r_data_new[start + k] = val
                
        self.r_msgs.data = r_data_new

    def compute_variable_to_check(self):
        """
        Update variable to check messages (q) and posteriors (M).
        q_{vc} = Lambda_v + sum_{c' != c} r_{c'v}
        """
        # Strategy:
        # 1. Calculate full sum for each variable (Posterior M_v)
        # 2. Subtract r_{cv} to get q_{vc}
        
        # Convert r_msgs to CSC to sum over columns efficiently
        r_csc = self.r_msgs.tocsc()
        
        # 1. Compute Priors with Memory
        # Lambda_v(t) = (1 - gamma_v) * Lambda_v(0) + gamma_v * M_v(t-1)
        if self.gamma0 is not None:
            # Vectorized update
            current_priors = (self.log_prior_ratios * (1 - self.memory_strengths) + 
                              self.posterior_ratios * self.memory_strengths)
        else:
            current_priors = self.log_prior_ratios
            
        # 2. Calculate new Posteriors (sum of all incoming r messages + prior)
        # r_csc.sum(axis=0) sums columns. Returns matrix of shape (1, num_vars)
        incoming_sum = np.array(r_csc.sum(axis=0)).flatten()
        new_posteriors = current_priors + incoming_sum
        
        # Update stored posteriors
        self.posterior_ratios = new_posteriors.copy()
        
        # 3. Update q messages
        # q_{vc} = new_posterior - r_{cv}
        # r_csc.data contains r_{cv} values. 
        # r_csc.indices contains row indices (checks).
        
        indptr = r_csc.indptr
        indices = r_csc.indices
        data = r_csc.data
        
        q_data_csc = np.zeros_like(data)
        
        for j in range(self.num_vars):
            start = indptr[j]
            end = indptr[j+1]
            if start == end:
                continue
                
            total = new_posteriors[j]
            # q_{vc} = total - r_{cv}
            q_data_csc[start:end] = total - data[start:end]
            
        # Create new CSC matrix for Q
        q_csc = csc_matrix((q_data_csc, indices, indptr), shape=self.check_matrix.shape)
        
        # Convert back to CSR to store
        self.q_msgs = q_csc.tocsr()

    def step(self, syndrome):
        self.compute_check_to_variable(syndrome)
        self.compute_variable_to_check()
        self.current_iteration += 1
        
    def get_hard_decision(self):
        # Negative LLR means likelihood of error is higher (in some conventions).
        # Standard convention: LLR = log(P(no error) / P(error))
        # So LLR < 0 implies Error.
        return (self.posterior_ratios <= 0).astype(np.uint8)

    def decode(self, syndrome):
        # Only initialize if it's the first run or if explicitly reset externally
        # But for standard use, we assume clean slate unless part of relay
        if self.current_iteration == 0:
            self.initialize_decoder()
        
        for _ in range(self.max_iter):
            self.step(syndrome)
            
            # Check convergence
            candidate = self.get_hard_decision()
            # H * candidate = syndrome ?
            calc_syndrome = (self.check_matrix @ candidate) % 2
            if np.array_equal(calc_syndrome, syndrome):
                self.decoding = candidate
                return True, candidate
                
        self.decoding = self.get_hard_decision()
        return False, self.decoding

class RelayDecoder:
    def __init__(self, check_matrix, error_priors, 
                 pre_iter=80, num_legs=300, iter_per_leg=60,
                 gamma_interval=(-0.24, 0.66), gamma0=0.1):
        """
        Relay-BP Decoder.
        
        Args:
            check_matrix: Parity check matrix (scipy sparse or dense).
            error_priors: Error probabilities.
            pre_iter: Iterations for the first leg.
            num_legs: Maximum number of relay legs.
            iter_per_leg: Iterations per subsequent leg.
            gamma_interval: (min, max) for random memory strength.
            gamma0: Initial memory strength for the first leg.
        """
        self.bp_decoder = MinSumBPDecoder(
            check_matrix, error_priors, 
            max_iter=pre_iter, 
            gamma0=gamma0
        )
        self.pre_iter = pre_iter
        self.num_legs = num_legs
        self.iter_per_leg = iter_per_leg
        self.gamma_interval = gamma_interval
        self.gamma0 = gamma0
        
        self.num_vars = self.bp_decoder.num_vars
        
    def _generate_gammas(self):
        """Generate random memory strengths from uniform distribution."""
        low, high = self.gamma_interval
        return np.random.uniform(low, high, size=self.num_vars)
        
    def decode(self, syndrome):
        """
        Run the Relay-BP decoding process.
        Returns:
            success (bool): Whether a valid correction was found.
            correction (np.ndarray): The best correction found (lowest weight).
        """
        # Track best solution
        best_correction = None
        min_weight = float('inf')
        
        # 1. Run First Leg (Leg 0)
        # Using gamma0 (usually small positive, e.g. 0.1)
        self.bp_decoder.max_iter = self.pre_iter
        self.bp_decoder.set_memory_strengths(np.full(self.num_vars, self.gamma0))
        self.bp_decoder.initialize_decoder()
        
        # We manually step to allow checking (though decode() does this too)
        # But we want to preserve state for next leg.
        success, correction = self.bp_decoder.decode(syndrome)
        
        if success:
            weight = np.sum(correction) # Hamming weight (simplification, ideally weighted by priors)
            # Paper uses weighted weight: sum(e_j * log((1-p)/p))
            # But simple Hamming is often sufficient for CSS unless p varies wildly.
            # Let's use the actual log prior weight.
            weight = np.sum(correction * self.bp_decoder.log_prior_ratios)
            
            if weight < min_weight:
                min_weight = weight
                best_correction = correction.copy()
                
            # If standard BP converges, we might just return?
            # Paper says: "The algorithm stops when R legs have executed or S solutions have been found"
            # If S=1, we stop here. Let's assume S=1 for now.
            return True, best_correction

        # 2. Relay Loop
        for leg in range(self.num_legs):
            # Update configuration for relay legs
            self.bp_decoder.max_iter = self.iter_per_leg
            
            # Generate new disordered memory strengths
            new_gammas = self._generate_gammas()
            self.bp_decoder.set_memory_strengths(new_gammas)
            
            # Reset messages but keep posteriors
            self.bp_decoder.reset_for_new_leg()
            
            # Run this leg
            success, correction = self.bp_decoder.decode(syndrome)
            
            if success:
                weight = np.sum(correction * self.bp_decoder.log_prior_ratios)
                if weight < min_weight:
                    min_weight = weight
                    best_correction = correction.copy()
                
                # We found a solution. 
                # If we want just one valid solution, we can return.
                # If we want the *best* solution, we might continue?
                # Usually stopping on first success is standard for speed.
                return True, best_correction
                
        # If we exhausted all legs
        if best_correction is not None:
            return True, best_correction
            
        return False, self.bp_decoder.decoding
